README.md

```diff
@@ -8,3 +8,103 @@ DI friendly to make it easier to test your code. You can test what happens on
 specific encode/decode errors without having to craft input that causes them.
 
 Just mock the Encoder and have it behave the way you want to test.
+
+## Encoder vs Trait
+
+There is no behavioural difference between the Encoder and the Trait.
+
+Use the Encoder version for any service class.  This let's you mock it and thus
+mock encode/decode issues without having to craft inputs that produce them.
+
+Use the Trait when you need the same functionality in a data class.  Eg imagine
+a `JsonClient` that returns a `JsonResponse` that wants to `decode` the payload.
+
+There's a variety of ways to handle this scenario.
+
+1.  Inject the encoder as a constructor arg after the others.
+
+```
+class JsonResponse
+{
+    public function __construct(
+        private readonly ResponseInterface $response,
+        private readonly JsonEncoder $json = new JsonEncoder,
+    ) {
+    }
+}
+```
+
+This works but that's a very heavy handed way to be able to mock the issue when
+in reality we're likely to be mocking JsonResponse anyway.
+
+2.  We can have the `JsonClient` do the `decode` and pass the result into our
+`JsonResponse`.
+
+```
+class JsonResponse
+{
+    public function __construct(
+        private readonly ResponseInterface $response,
+        private readonly array $data,
+    ) {
+    }
+}
+```
+
+This produces the `JsonException` at the right point.  We can mock the
+`JsonEncoder` that `JsonClient` uses etc.  The price we pay is that we always
+`decode` the payload even if it never gets used.
+
+3.  The chosen approach.  We make the functionality available as a `Trait` and
+assume that we'll mock that class to throw our chosen `JsonException`.
+
+```
+class JsonResponse
+{
+    use JsonTrait;
+
+    public function __construct(
+        private readonly ResponseInterface $response,
+    ) {
+    }
+}
+```
```

Comment:
Let's imagine we have `JsonResponse::getData(): array();` method to return decoded response and we need to test it. When we use the trait we need to mock `ResponseInterface` and it should return encoded content so it seems we we'll test the trait multiple times. Smth like:

```
$mock = self::createMock(ResponseInteface::class);
$mock->method('getContent')->willReturn(json_encode($data));

assert::same($data, $jsonResponse->getData());
```

When we pass it in constructor, we just need:

```
$this->json->method('decode')->willReturn($data);
assert::same($data, $jsonResponse->getData());
```

Reply:
Cool, thanks.  That's the same conclusion I came to while writing that explanation.

Compared to these other factors, always decoding the body seems kinda irrelevant and I now prefer 2.  Which means this whole PR is irrelevant.